### ISSUE TITLE
A random split function that return the datasets following specific target (label) distribution.

### ISSUE THREAD

**Issue by louis2889184** [pytorch/pytorch#46711]
## 🚀 Feature
A random split function that return the datasets following specific target (label) distribution.

## Motivation

Sometimes, we might want to control the data distribution while splitting the data. For example, uniformly distributed labeled data in semi-supervised learning (10 classes and each class has 100 data) and constructing imbalance/distribution-shift data.

## Pitch

```python
# Current random_split function
torch.utils.data.random_split(dataset, lengths, generator=<torch._C.Generator object>)

# Possible distributional_split function
torch.utils.data.distributional_split(dataset, lengths, get_label, target_distribution, generator=<torch._C.Generator object>):
"""
Args:
    dataset: Dataset in PyTorch
    lengths: the lengths of split datasets (we could restrict to only two datasets here, A and B)
    get_label: function to return label from dataset
    target_distribution: label distributions of the first split datasets (that is A).
Return:
    the splitted datasets, where the first one follows the target distribution.
"""
```

## Additional context

If it is of interested, I am happy to help with this.
